Activity feed: Surface team_id and team_name in policy and query activities

drvcodenta opened the pull request:
fix #27689
# Checklist for submitter

- [x] Changes file added for user-visible changes in `changes/`
- [x] Input data is properly validated, `SELECT *` is avoided, SQL injection is prevented (using placeholders for values in statements)
- [x] Added/updated automated tests
- [x] A detailed QA plan exists on the associated ticket (added in the comment)
- [x] Manual QA for all new/changed functionality

output ss
![Screenshot 2025-04-20 100928](https://github.com/user-attachments/assets/b62c1370-18e6-4d0f-a8ab-3a3d2d2084c7)
![Screenshot 2025-04-20 101101](https://github.com/user-attachments/assets/4f052242-d4c0-4027-afef-37f8d29991fc)

Noticed that editing a global policy doesn't generate an edited_policy activity log in my local environment, while it does in others. Investigated the gap and shared findings in the issue thread for further clarification(in https://github.com/fleetdm/fleet/issues/27689)

sgress454 commented on server/service/queries.go:
```diff
@@ -263,6 +263,17 @@ func createQueryEndpoint(ctx context.Context, request interface{}, svc fleet.Ser
 	return createQueryResponse{query, nil}, nil
 }
 
+func (svc *Service) getTeamName(ctx context.Context, teamID *uint) (*string, error) {
+	if teamID == nil {
+		return nil, nil
+	}
+	team, err := svc.ds.Team(ctx, *teamID)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "fetching team name")
+	}
+	return &team.Name, nil
+}
+
```
Rather than implementing a new method just to get the team name, you can utilize the `EnterpriseOverrides` pattern to conditionally retrieve the team name if we're on a licensed tier, e.g.

```go
var teamName *string
if query.TeamID != nil {
	team, err := svc.EnterpriseOverrides.TeamByIDOrName(ctx, query.TeamID, nil)
	if err != nil {
		return nil, err
	}
	teamName = &team.Name
}
```

sgress454 reviewed (commented):
Nice work @drvcodenta!  Just a couple of issues to address:

* This fails when working with policies on the "No Team" team, since it's not a "real" team in the database.  You'll need a special case for when the supplied team ID is `0`.
* As it stands, any previous query/policy activity will be shown as "global" even if it was for a team, which would be incorrect.  One possible solve for this would be to set the team ID in the activity to -1 rather than `nil`, and check for that in the front end. 

I think some of the test failures you're seeing may be relieved by merging `main` into your branch!

RachelElysia commented on frontend/pages/DashboardPage/cards/ActivityFeed/GlobalActivityItem/GlobalActivityItem.tsx:
```diff
@@ -1143,6 +1143,96 @@ const TAGGED_TEMPLATES = {
+            on the <b>{activity.details.team_name}</b> team
+          </>
+        ) : (
+          "globally"
```
Design cc: @eugkuo --

Can you confirm copy text for adding/editing/deleting queries/policies that are global?

eugkuo replied:
@RachelElysia You mean when something is done for all teams? And whether indicate that? I just created a policy for all teams and this is the activity message:
![Screenshot 2025-04-23 at 17 09 34](https://github.com/user-attachments/assets/61af8e69-1542-482a-856c-f6729ff9bb13)

So I guess we just don't show a team.

RachelElysia commented on frontend/pages/DashboardPage/cards/ActivityFeed/GlobalActivityItem/GlobalActivityItem.tsx:
```diff
@@ -1425,6 +1515,24 @@ const getDetail = (activity: IActivity, isPremiumTier: boolean) => {
+    }
+    case ActivityType.DeletedPolicy: {
+      return TAGGED_TEMPLATES.deletedPolicy(activity);
+    }
```
While in this code, I noticed there's an `ActivityType.DeletedMultipleSavedQuery`. To be consistent, I think we would want to add team information to this activity as well. WDYT @drvcodenta / design @eugkuo

drvcodenta replied:
Yeah, I think that’s a good idea.
We can include the team information when deleting multiple saved queries as well — makes sense to keep it consistent. I’ll make the necessary changes.

RachelElysia reviewed (commented):
Just went through the FE, things look great @drvcodenta ! Just confirming with design for the copy text and if we want to add team information to the bulk delete queries, I think it's safe to assume so.

@mostlikelee I can approve the FE, but we need BE reviewers eyes on this as well. Will bring up tomorrow in standup.

RachelElysia commented:
Actually, looks like @sgress454 is on top of reviewing BE, thanks @sgress454

drvcodenta commented:
Hey team 👋

Apologies for the delay — I’ve been really unwell over the past week and wasn’t able to follow up here.
I’m feeling better now and will start addressing the feedback right away. Thanks for your patience!

sgress454 reviewed (commented):
This is looking good @drvcodenta!  Some minor issues around copy, backwards compatibility and linting, but this is very close.  Thanks for working through this, looking forward to getting it across the finish line.

drvcodenta commented:
Thanks for the clarification! I've added omitempty to the new team_name fields, updated copy text, and comments to bypass linting suggested. Let me know if there's anything else you'd like adjusted.

lukeheath commented:
@sgress454 Just a reminder that this is ready for re-review.

sgress454 reviewed (commented):
Sorry for the delay, this is looking good!

The last bit is to address the failing tests.  The ones in `integration_core_test` probably need to be moved to `integration_enterprise_test`; it looks like they're failing due to `EnterpriseOverrides` not being set, which makes sense since the `integration_core_test` doesn't test premium features.

Other tests look like they need their expected output updated to include `team_name`.

There's a failing SCEP test that seems unrelated; try merging `main` back to your branch and see if it resolves itself.

Let me know if you have any struggles getting the tests to pass!

lukeheath commented:
Converting this PR to draft until it is ready to merge to avoid impacting our weekly PR open time KPIs.

sgress454 reviewed (dismissed):
Thanks for your hard work on this @drvcodenta!

rachaelshaw commented on docs/Contributing/reference/audit-logs.md:
```diff
@@ -207,12 +231,16 @@ Generated when deleting multiple saved queries.
 This activity contains the following fields:
 - "query_ids": list of IDs of the deleted saved queries.
+- "team_id": the ID of the team the queries belonged to.
+- "team_name": the name of the team the queries belonged to, "Global" if global.
```
I'd rather leave `team_name` empty for global queries. Rather than introducing a new pattern for indicating whether something is global or team-level, how about we use a `global` boolean like we do in the `edited_agent_options` activity?

drvcodenta replied:
Yes, for global activities (query, policy), the current implementation sets team_id to -1 and leaves team_name as nil in both the struct and service layer.

On the frontend, we handle these activities with a conditional check: if team_id is -1 or if team_name is present (which would be the case when a valid team_id exists). For older activities with no associated team, team_name is an empty string.

Regarding the DeletedMultipleSavedQuery activity, the backend implementation follows the same pattern. However, on the frontend, the check currently only verifies whether team_name is present and accordingly renders either the team name or "globally." In hindsight, this may have been an oversight on my part, but the auto-generated activity documentation does reflect the frontend behavior accurately.

For all other activities—such as create, edit, or delete actions for queries and policies—the system correctly omits the team_name string when no team information is available and uses -1 for global activities.

<img width="715" height="694" alt="image" src="https://github.com/user-attachments/assets/cafb54a4-b8c9-4a88-b1f5-148ba05c40a8" />
